{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "V4o6xsVQ7oHy"
   },
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import xml.etree.ElementTree as etree\n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "khQ_YuTm8WRu"
   },
   "outputs": [],
   "source": [
    "# Set URL for XML file with all BioSamples associated with project\n",
    "from urllib.request import urlopen\n",
    "url = 'https://www.ncbi.nlm.nih.gov/portal/utils/file_backend.cgi?Db=biosample&HistoryId=NCID_1_125617756_130.14.18.48_5555_1587647699_834710852_0MetA0_S_HStore&QueryKey=8&Sort=&Filter=all&CompleteResultCount=683&Mode=file&View=fullxml&p$l=Email&portalSnapshot=%2Fprojects%2FBioSample%2Fbiosample%401.33&BaseUrl=&PortName=live&RootTag=BioSampleSet&FileName=&ContentType=xml'\n",
    "xml_file = urlopen(url)\n",
    "\n",
    "tree = etree.parse(xml_file)\n",
    "root = tree.getroot()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "HyppWYZp_n38"
   },
   "outputs": [],
   "source": [
    "# Append list of each all attributes for each sample within XML\n",
    "all_samples = []\n",
    "for child in root:\n",
    "  samples = []\n",
    "  for child in child:\n",
    "    attribute_element = child.findall('Attribute')\n",
    "    if attribute_element is not None:\n",
    "      elements = []\n",
    "      for element in attribute_element:\n",
    "        individual = [element.items()[0][1], element.text]\n",
    "        elements.append(individual)\n",
    "      samples.append(elements)\n",
    "  all_samples.append(samples)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 34
    },
    "colab_type": "code",
    "executionInfo": {
     "elapsed": 400,
     "status": "ok",
     "timestamp": 1587512271074,
     "user": {
      "displayName": "Ryan Martin",
      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14GibETOFeizGOAk3zCp7KIrGXJAo5ZHDuhOFhM1Khw=s64",
      "userId": "18244151384388935977"
     },
     "user_tz": 240
    },
    "id": "UPYbVuTcdMpw",
    "outputId": "2869d67d-eaa5-47c3-b4bc-b2cba35f48b8"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "51"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Replace all spaces with underscores to remove duplicates\n",
    "for i in range(len(all_samples)):\n",
    "  for num in range(len(all_samples[i][5])):\n",
    "    all_samples[i][5][num][0] = all_samples[i][5][num][0].replace(' ', '_')\n",
    "\n",
    "# Full list of Attributes across all sample \n",
    "features = []\n",
    "for i in range(len(all_samples)):\n",
    "  for x, y in all_samples[i][5]:\n",
    "    features.append(x)\n",
    "features = set(features)\n",
    "len(features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 136
    },
    "colab_type": "code",
    "executionInfo": {
     "elapsed": 7072,
     "status": "ok",
     "timestamp": 1587512278670,
     "user": {
      "displayName": "Ryan Martin",
      "photoUrl": "https://lh3.googleusercontent.com/a-/AOh14GibETOFeizGOAk3zCp7KIrGXJAo5ZHDuhOFhM1Khw=s64",
      "userId": "18244151384388935977"
     },
     "user_tz": 240
    },
    "id": "Hi2gfpzxpMQt",
    "outputId": "69a2067e-1483-456f-9cd3-3cd1eb23201f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "100\n",
      "200\n",
      "300\n",
      "400\n",
      "500\n",
      "600\n"
     ]
    }
   ],
   "source": [
    "#Left join each sample by feature to create dataframe with all samples\n",
    "\n",
    "df = pd.DataFrame(features, columns = ['features'])\n",
    "df.set_index(['features'], inplace = True)\n",
    "              \n",
    "for i in range(len(all_samples)):\n",
    "    temp = pd.DataFrame(all_samples[i][5], columns = ['features', str('Sample' + '_' + str(i))])\n",
    "    temp.set_index(['features'], inplace = True)\n",
    "    df = df.join(temp, how = 'left')\n",
    "    if i % 100 == 0:\n",
    "      print(str(i))\n",
    "\n",
    "# Transpose so each row is a subject and each column is sample\n",
    "df = df.T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>features</th>\n",
       "      <th>INSDC_status</th>\n",
       "      <th>fecal_date</th>\n",
       "      <th>study_center</th>\n",
       "      <th>INSDC_center_alias</th>\n",
       "      <th>experiment_title</th>\n",
       "      <th>physical_specimen_remaining</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>year_diagnosed</th>\n",
       "      <th>samp_size</th>\n",
       "      <th>uc_extent</th>\n",
       "      <th>...</th>\n",
       "      <th>timepoint</th>\n",
       "      <th>Sample_Name</th>\n",
       "      <th>geo_loc_name</th>\n",
       "      <th>dna_extracted</th>\n",
       "      <th>sample_type</th>\n",
       "      <th>bmi</th>\n",
       "      <th>cd_location</th>\n",
       "      <th>longitude</th>\n",
       "      <th>ENA_checklist</th>\n",
       "      <th>body_product</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>Sample_0</td>\n",
       "      <td>public</td>\n",
       "      <td>23-Jun-09</td>\n",
       "      <td>LBNL</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Jansson_Lamendella_Crohns</td>\n",
       "      <td>false</td>\n",
       "      <td>1629:739</td>\n",
       "      <td>1966</td>\n",
       "      <td>.1,g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ERS1464576</td>\n",
       "      <td>Sweden</td>\n",
       "      <td>true</td>\n",
       "      <td>stool</td>\n",
       "      <td>27</td>\n",
       "      <td>Ileal (L1)</td>\n",
       "      <td>18.644</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>UBERON:feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sample_1</td>\n",
       "      <td>public</td>\n",
       "      <td>22-Apr-10</td>\n",
       "      <td>LBNL</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Jansson_Lamendella_Crohns</td>\n",
       "      <td>false</td>\n",
       "      <td>1629:1004</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.1,g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ERS1464575</td>\n",
       "      <td>Sweden</td>\n",
       "      <td>true</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.644</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>UBERON:feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sample_2</td>\n",
       "      <td>public</td>\n",
       "      <td>27-Dec-10</td>\n",
       "      <td>LBNL</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Jansson_Lamendella_Crohns</td>\n",
       "      <td>false</td>\n",
       "      <td>1629:908</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.1,g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>ERS1464574</td>\n",
       "      <td>Sweden</td>\n",
       "      <td>true</td>\n",
       "      <td>stool</td>\n",
       "      <td>22</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.644</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>UBERON:feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sample_3</td>\n",
       "      <td>public</td>\n",
       "      <td>14-Jul-09</td>\n",
       "      <td>LBNL</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Jansson_Lamendella_Crohns</td>\n",
       "      <td>false</td>\n",
       "      <td>1629:1015</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.1,g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ERS1464573</td>\n",
       "      <td>Sweden</td>\n",
       "      <td>true</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.644</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>UBERON:feces</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>Sample_4</td>\n",
       "      <td>public</td>\n",
       "      <td>26-May-10</td>\n",
       "      <td>LBNL</td>\n",
       "      <td>UCSDMI</td>\n",
       "      <td>Jansson_Lamendella_Crohns</td>\n",
       "      <td>false</td>\n",
       "      <td>1629:1018</td>\n",
       "      <td>NaN</td>\n",
       "      <td>.1,g</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>ERS1464572</td>\n",
       "      <td>Sweden</td>\n",
       "      <td>true</td>\n",
       "      <td>stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18.644</td>\n",
       "      <td>ERC000011</td>\n",
       "      <td>UBERON:feces</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 51 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "features INSDC_status fecal_date study_center INSDC_center_alias  \\\n",
       "Sample_0       public  23-Jun-09         LBNL             UCSDMI   \n",
       "Sample_1       public  22-Apr-10         LBNL             UCSDMI   \n",
       "Sample_2       public  27-Dec-10         LBNL             UCSDMI   \n",
       "Sample_3       public  14-Jul-09         LBNL             UCSDMI   \n",
       "Sample_4       public  26-May-10         LBNL             UCSDMI   \n",
       "\n",
       "features           experiment_title physical_specimen_remaining  \\\n",
       "Sample_0  Jansson_Lamendella_Crohns                       false   \n",
       "Sample_1  Jansson_Lamendella_Crohns                       false   \n",
       "Sample_2  Jansson_Lamendella_Crohns                       false   \n",
       "Sample_3  Jansson_Lamendella_Crohns                       false   \n",
       "Sample_4  Jansson_Lamendella_Crohns                       false   \n",
       "\n",
       "features host_subject_id year_diagnosed samp_size uc_extent  ... timepoint  \\\n",
       "Sample_0        1629:739           1966      .1,g       NaN  ...         2   \n",
       "Sample_1       1629:1004            NaN      .1,g       NaN  ...         2   \n",
       "Sample_2        1629:908            NaN      .1,g       NaN  ...         7   \n",
       "Sample_3       1629:1015            NaN      .1,g       NaN  ...         2   \n",
       "Sample_4       1629:1018            NaN      .1,g       NaN  ...         2   \n",
       "\n",
       "features Sample_Name geo_loc_name dna_extracted sample_type  bmi cd_location  \\\n",
       "Sample_0  ERS1464576       Sweden          true       stool   27  Ileal (L1)   \n",
       "Sample_1  ERS1464575       Sweden          true       stool  NaN         NaN   \n",
       "Sample_2  ERS1464574       Sweden          true       stool   22         NaN   \n",
       "Sample_3  ERS1464573       Sweden          true       stool  NaN         NaN   \n",
       "Sample_4  ERS1464572       Sweden          true       stool  NaN         NaN   \n",
       "\n",
       "features longitude ENA_checklist  body_product  \n",
       "Sample_0    18.644     ERC000011  UBERON:feces  \n",
       "Sample_1    18.644     ERC000011  UBERON:feces  \n",
       "Sample_2    18.644     ERC000011  UBERON:feces  \n",
       "Sample_3    18.644     ERC000011  UBERON:feces  \n",
       "Sample_4    18.644     ERC000011  UBERON:feces  \n",
       "\n",
       "[5 rows x 51 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "orKIraBVfnjG"
   },
   "outputs": [],
   "source": [
    "df.reset_index(inplace = True)\n",
    "df.rename(columns = {'index' : 'samples'}, inplace= True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "_ewSi3r78spo"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>study_accession</th>\n",
       "      <th>sample_accession</th>\n",
       "      <th>secondary_sample_accession</th>\n",
       "      <th>experiment_accession</th>\n",
       "      <th>run_accession</th>\n",
       "      <th>tax_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>library_layout</th>\n",
       "      <th>fastq_ftp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>PRJEB18471</td>\n",
       "      <td>SAMEA19334668</td>\n",
       "      <td>ERS1463894</td>\n",
       "      <td>ERX1815536</td>\n",
       "      <td>ERR1746339</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/009/ERR174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>PRJEB18471</td>\n",
       "      <td>SAMEA19335418</td>\n",
       "      <td>ERS1463895</td>\n",
       "      <td>ERX1815537</td>\n",
       "      <td>ERR1746340</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/000/ERR174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>PRJEB18471</td>\n",
       "      <td>SAMEA19336168</td>\n",
       "      <td>ERS1463896</td>\n",
       "      <td>ERX1815538</td>\n",
       "      <td>ERR1746341</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/001/ERR174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PRJEB18471</td>\n",
       "      <td>SAMEA19336918</td>\n",
       "      <td>ERS1463897</td>\n",
       "      <td>ERX1815539</td>\n",
       "      <td>ERR1746342</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/002/ERR174...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>PRJEB18471</td>\n",
       "      <td>SAMEA19337668</td>\n",
       "      <td>ERS1463898</td>\n",
       "      <td>ERX1815540</td>\n",
       "      <td>ERR1746343</td>\n",
       "      <td>408170</td>\n",
       "      <td>human gut metagenome</td>\n",
       "      <td>Illumina HiSeq 2000</td>\n",
       "      <td>SINGLE</td>\n",
       "      <td>ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/003/ERR174...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  study_accession sample_accession secondary_sample_accession  \\\n",
       "0      PRJEB18471    SAMEA19334668                 ERS1463894   \n",
       "1      PRJEB18471    SAMEA19335418                 ERS1463895   \n",
       "2      PRJEB18471    SAMEA19336168                 ERS1463896   \n",
       "3      PRJEB18471    SAMEA19336918                 ERS1463897   \n",
       "4      PRJEB18471    SAMEA19337668                 ERS1463898   \n",
       "\n",
       "  experiment_accession run_accession  tax_id       scientific_name  \\\n",
       "0           ERX1815536    ERR1746339  408170  human gut metagenome   \n",
       "1           ERX1815537    ERR1746340  408170  human gut metagenome   \n",
       "2           ERX1815538    ERR1746341  408170  human gut metagenome   \n",
       "3           ERX1815539    ERR1746342  408170  human gut metagenome   \n",
       "4           ERX1815540    ERR1746343  408170  human gut metagenome   \n",
       "\n",
       "      instrument_model library_layout  \\\n",
       "0  Illumina HiSeq 2000         SINGLE   \n",
       "1  Illumina HiSeq 2000         SINGLE   \n",
       "2  Illumina HiSeq 2000         SINGLE   \n",
       "3  Illumina HiSeq 2000         SINGLE   \n",
       "4  Illumina HiSeq 2000         SINGLE   \n",
       "\n",
       "                                           fastq_ftp  \n",
       "0  ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/009/ERR174...  \n",
       "1  ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/000/ERR174...  \n",
       "2  ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/001/ERR174...  \n",
       "3  ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/002/ERR174...  \n",
       "4  ftp.sra.ebi.ac.uk/vol1/fastq/ERR174/003/ERR174...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Build link to project on EBI and download sample data\n",
    "\n",
    "base = 'https://www.ebi.ac.uk/ena/data/warehouse/filereport'\n",
    "project = 'PRJEB18471'\n",
    "result = 'read_run'\n",
    "fields = 'study_accession,sample_accession,secondary_sample_accession,experiment_accession,run_accession,tax_id,scientific_name,instrument_model,library_layout,fastq_ftp'\n",
    "download = 'txt'\n",
    "\n",
    "url = base + '?accession=' + project + '&result=' + result + '&fields=' + fields + '&download=' + download\n",
    "\n",
    "download_links = pd.read_csv(url, sep = '\\t')\n",
    "\n",
    "# Check download is correct\n",
    "download_links.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "colab": {},
    "colab_type": "code",
    "id": "8vXQgfSBZnJ-"
   },
   "outputs": [],
   "source": [
    "# Merge data parsed from XML with ftp links for sequencing data from EBI\n",
    "df = df.merge(download_links, how = 'left', left_on = 'SRA_accession', right_on = 'secondary_sample_accession')"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyMh94mQAvSiBvz8UNTl2ZwD",
   "machine_shape": "hm",
   "name": "FindIBDSamples.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
